Prior auth

trafficpest opened the pull request:
Prior Auth bug fixes and Features
fixed date type hint bug on a funtion when adding new auth

changed CPTs to CPT since you can only add one singular at a time

Removed duplicates rows in global report

added pie graph of % of all auths used on global report

Added bar graph and percent in tables of patient auths and global
auth reports to easily see how many auths you have remaining.
To know when to reauth quickly
It is color coded green 67-100%, yellow 34-66%, red 0-33%

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/index.php:
```diff
@@ -161,6 +162,29 @@ function refreshme() {
+                            $barColor = '#dc3545'; // Red: Very low
+                        }
+
+                        print TABLE_TD;
```
echo is probably marginally better than print for these.

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/index.php:
```diff
@@ -161,6 +162,29 @@ function refreshme() {
                         print TABLE_TD . text($iter['cpt']);
+
+                        $initialUnits = (int)$iter['init_units'];
+
+                        if ($initialUnits > 0) {
+                            $percentRemaining = round(($remaining / $initialUnits) * 100);
+                        } else {
+                            $percentRemaining = 0;
+                        }
+                        $barColor = '#4CAF50'; // Green: Full/Plenty remaining
+                        if ($percentRemaining <= 66) {
+                            $barColor = '#ffc107'; // Yellow: Getting low
+                        }
+                        if ($percentRemaining <= 33) {
+                            $barColor = '#dc3545'; // Red: Very low
+                        }
```
Consider using a match expression here. Note that if $percentRemaining <= 33 it's also <= 66, so the assignment is done three times in that case.

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -26,6 +30,7 @@
           content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
     <meta http-equiv="X-UA-Compatible" content="ie=edge">
     <title><?php echo xlt("List Exising Prior Auths Report"); ?></title>
+<script src="https://cdn.jsdelivr.net/npm/chart.js@4.4.0/dist/chart.umd.min.js"></script>
```
Please check indentation.

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -16,6 +16,10 @@
 $data = new AuthorizationService();
 $patients = $data->listPatientAuths();
 
+$hide_expired = isset($_GET['hide_expired']) && $_GET['hide_expired'] == '1';
```
```suggestion
$hide_expired = ($_GET['hide_expired'] ?? '0') === '1';
```

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -92,21 +121,52 @@ function openNewTopWindow(newpid) {
                     print "<td>" . text($iter['end_date']) . "</td>";
-                    if (($iter['end_date'] < date('Y-m-d')) && ($iter['end_date'] !== '0000-00-00') && !empty($iter['auth_num'])) {
+
+                    if (($iter['end_date'] < date('Y-m-d')) && ($iter['end_date'] !== '0000-00-00') && (!empty($iter['auth_num']))) {
```
Consider assigning the $iter['end_date'] check earlier to avoid duplication.

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -92,21 +121,52 @@ function openNewTopWindow(newpid) {
+                        $initialUnits = (int)$iter['init_units'];
+                        print "<td>" . text($initialUnits) . "</td>";
+
+                        $usedUnits = $initialUnits - $numbers;
+                        $unitCount = $initialUnits - $numbers;
```
```suggestion
                        $usedUnits = $initialUnits - $numbers;
                        $unitCount = $usedUnits;
```

kojiromike commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -92,21 +121,52 @@ function openNewTopWindow(newpid) {
+                            $percentRemaining = 0;
+                        }
+
+                        $barColor = '#4CAF50'; // Green: Full/Plenty remaining
+                        if ($percentRemaining <= 66) {
+                            $barColor = '#ffc107'; // Yellow: Getting low
+                        }
+                        if ($percentRemaining <= 33) {
+                            $barColor = '#dc3545'; // Red: Empty
+                        }
```
```suggestion
                        $barColor = match(true) {
                            ($percentRemaining <= 33) => '#dc3545'; // Red: Empty
                            ($percentRemaining <= 66) => '#ffc107'; // Yellow: Getting low
                            default => '#4CAF50'; // Green: Full/Plenty remaining
                        };
```

You might also consider assigning the colors to variables or even an enum so the "empty/getting low/full" meaning can be gotten from the names.

kojiromike reviewed (commented):
Nice! I left some suggestions for simplifying some code and avoiding duplication.

stephenwaite commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/index.php:
```diff
@@ -150,22 +151,44 @@ function refreshme() {
+
+                        echo TABLE_TD;
+                        echo "<div style='background-color:#eee; height:20px; width:100px; border:1px solid #ccc; position:relative;'>";
+                        echo "<div style='background-color:{$barColor}; height:100%; width:{$percentRemaining}%; position:absolute;'></div>";
```
hi @trafficpest, have to escape these variables:
```
                        echo "<div style='background-color:{" . attr($barColor) . "}; height:100%; width:{" . attr($percentRemaining) . "}%; position:absolute;'></div>";
```
there's at least one other spot in this PR

trafficpest replied:
ok ill check now

trafficpest replied:
ok did see two. I dont think i missed any more

kojiromike commented:
Is there an issue for this PR? We will need one, IIUC.

trafficpest commented:
> Is there an issue for this PR? We will need one, IIUC.

No, sorry. My sister-in-law had a few bugs and wanted some more features, then I thought I would share. I will make one in a bit. You guys are so fancy for me, lol.

I'm learning how you like things as I go. Thank you again

stephenwaite commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -73,41 +98,75 @@ function openNewTopWindow(newpid) {
+                        echo "<td>";
+                        echo "<div style='background-color:#eee; height:20px; width:150px; border:1px solid #ccc; position:relative;'>";
+                        echo "<div style='background-color:" . attr($barColor) . "; height:100%; width:" . attr($percentRemaining) . "%; position:absolute;'></div>";
+                        echo "<div style='position:absolute; top:0; width:100%; text-align:center; line-height:20px; color:#000; font-weight:bold; font-size:12px;'>{$percentRemaining}%</div>";
```
and here :)

stephenwaite commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/reports/list_report.php:
```diff
@@ -120,6 +179,49 @@ function openNewTopWindow(newpid) {
         &copy; <?php echo date('Y') . " Juggernaut Systems Express" ?>
     </div>
+   <script>
+    const total_initial = <?php echo $total_initial_units; ?>;
```
these have to be `js_escaped()`

adunsulag commented:
@juggernautsei can you take a look over this, looks like a good addition, but want to make sure functionality wise it will still work the way your other clients are using it.

trafficpest commented on interface/modules/custom_modules/oe-module-prior-authorizations/public/index.php:
```diff
@@ -150,22 +151,44 @@ function refreshme() {
+                        echo TABLE_TD;
+                        echo "<div style='background-color:#eee; height:20px; width:100px; border:1px solid #ccc; position:relative;'>";
+                        echo "<div style='background-color:" . attr($barColor) . "; height:100%; width:" . attr($percentRemaining) . "%; position:absolute;'></div>";
+                        echo "<div style='position:absolute; top:0; width:100%; text-align:center; line-height:20px; color:#000; font-weight:bold; font-size:12px;'>{$percentRemaining}%</div>";
```
im so ignorant on these things I left $percentRemaining I thought since it returns a int it would be fine. thanks for teaching me

trafficpest commented:
@juggernautsei 
Note: the SQL count incorporates auth code + billed units of service code + service date to get an accurate count. My sister-in-law gets auth codes that include a date range and more than one service code each with different billed units. So now you enter amount for each code with auth. Thank you for the module, it was really helpful!

trafficpest commented:
This should be good to go now. I haven't tested since all these minor commits but I think it should be good.

juggernautsei commented:
I loaded the module and found in the index.php and the list_report.php needed to add the require autoloader in both files then it works. 
@sjpadgett  I noticed that there is a cdn reference. Is this permissible?

juggernautsei commented:
There is so much I would restructure about this module now. Hopefully, when we find an API partner. I will do the restructuring of this module to bring it up to date.

trafficpest commented:
> I loaded the module and found in the index.php and the list_report.php needed to add the require autoloader in both files then it works. @sjpadgett I noticed that there is a cdn reference. Is this permissible?

@kojiromike Why does the module need the vendor autoload still? I thought core app will include registered modules autoloads? Should the include be there or not?

juggernautsei commented:
@sjpadgett, I have already pulled down the module. Again, that is why I want to refactor this module to take advantage of the advances in the module ecosystem.  

When I first designed the module. I talked with Stephen briefly about it and the Patient Menu. Again, since there was no way at the time of injecting in the that menu. We opted to do the replacement which added the Prior Auth to the patient menu. Now, I should be able to expand the injection process to the patient menu. 

The other reason, I personally us an autoloader is sometimes I use PHP SlimFramwork in my modules. I like my modules having its own router. It gives me clean URL and decouples stuff like you using Phreezy. A lot of this stuff is just a matterr of preferences in my opinion. Of course we have to bow to the leaders of the gang of coders. There are still things I can't see yet but I will get there in time.  

So, I will refactor the openemrbootstrap.php file to include the class loader. Create a Bootstrap.php class in the /src and put in a really basic router. I am going to add and interface for the different prior auth providers or it may be better to use and abstract class. I will figure it out when I get closer to having a provider to work with.

sjpadgett commented:
@sjpadgett That's fine @juggernautsei it's your module and I'm really happy you contributed it. Can't wait to see outcome

sjpadgett commented:
@trafficpest @juggernautsei 
Sherwin, I believe you should be able to push your changes to a branch in the authors repo where he reviews and merges into his PR repo.
Once that is done we can merge to core.

trafficpest commented:
> @sjpadgett I will work that out. I don't have access to push changes to @trafficpest repo. I am stuck at this point
> 
@juggernautsei, sorry for the delay. I have sent an invite to you to be a collaborator and push code to my repo.
In the end what did you do with the module autoload? rewrite the bootstrap or just return the require in file?

Side note:
I have bad news for you guys: I got my wife's therapy practice to switch to OpenEMR after seeing it at my sister-in-law's ABA practice. Now you'll probably be dealing with me every once in a while, haha.

She already has me busy training employees at different locations, dealing with data migration, and, on top of that, I'm tasked with writing a Braintree payment integration module (her processor) that takes copays automatically on check-in. I'm tying it to arrival on the calendar, as I couldn't find an event for encounter.

trafficpest commented:
@juggernautsei, I see you're adding more than one CPT code in the string, and the way I changed the SQL to match the CPT and billed units, that would fail. I changed the label text from the plural CPTs to the singular CPT. Most authorizations have different units permitted for each service code, so the provider needs to add each one independently. The form could be changed, or the match condition could be changed from '=' to 'IN' in the SQL.

<img width="934" height="242" alt="Screenshot From 2025-10-07 23-16-59" src="https://github.com/user-attachments/assets/299a1201-5f20-4d08-8819-7f81cf4986a1" />

juggernautsei commented:
@trafficpest , I know one of my clients puts multiple CPTs in the field. For them to use the system as you have designed it. They would have to put in the same auth multiple times for each CPT?  We need to add an explainer on how to use the feature. So, people don't have to guess how the system is set up. I looked for the invite. I have not seen the invite as of yet. My email is sherwingaddis @ gmail.com.
